src/miner.cpp

```diff
@@ -121,15 +121,76 @@ uint64_t BlockAssembler::reserveBlockSize(const CScript &scriptPubKeyIn)
     nHeaderSize += 5; // tx count varint - 5 bytes is enough for 4 billion txs; 3 bytes for 65535 txs
 
     // This serializes with output value, a fixed-length 8 byte field, of zero and height, a serialized CScript
-    // signed integer taking up 4 bytes for heights 32768-8388607 (around the year 2167) after which it will use 5.
-    nCoinbaseSize = ::GetSerializeSize(coinbaseTx(scriptPubKeyIn, 400000, 0), SER_NETWORK, PROTOCOL_VERSION);
+    // signed integer taking up 4 bytes for heights 32768-8388607 (around the year 2167) after which it will use 5
+    nCoinbaseSize =
+        ::GetSerializeSize(coinbaseTx(scriptPubKeyIn, 400000, 0, coinbaseSize), SER_NETWORK, PROTOCOL_VERSION);
 
-    // BU Miners take the block we give them, wipe away our coinbase and add their own.
-    // So if their reserve choice is bigger then our coinbase then use that.
-    return nHeaderSize + std::max(nCoinbaseSize, coinbaseReserve.Value());
+    if (coinbaseSize < 0) // Explicit size of coinbase has not been requested
+    {
+        // BU Miners take the block we give them, wipe away our coinbase and add their own.
+        // So if their reserve choice is bigger then our coinbase then use that.
+        nCoinbaseSize = std::max(nCoinbaseSize, coinbaseReserve.Value());
+    }
+
+    return nHeaderSize + nCoinbaseSize;
+}
+
+
+/** Create a coinbase of specified size. Size must be multiple of 4*/
+static void MkFixedSizeCoinbaseTx(CMutableTransaction &tx, int64_t coinbaseSize)
+{
+    // Generates a bunch of bytes -Not a valid script.
+    // Removes any existing Tx(s)
+    int padsize = 0;
+    const int PAD76 = 76;
+    const int PAD78 = 78;
+    int maxpad = MAX_SCRIPT_SIZE; // Note: MAX_SCRIPT_SIZE --bytes are put into one script.
+
+    if (tx.vout.size() > 1)
+        LOGA("Discarding more than one transaction: %d.", tx.vout.size());
+
+    tx.vout.resize(1);
+    tx.vout[0].scriptPubKey.clear();
+    tx.vout[0].nValue = 0;
+
+    // Arg coinbaseSize 329 and 330 cause getminingcandidate() to generate 329+1, 330+1 size coinbase tx.
+    // By only accepting multiples of 4 the above is mitigated.
+    if (coinbaseSize % 4 != 0)
```

Comment:
Won't this be true for some negative values of `coinbaseSize` as well? I would suggest to use 0 for signalling "auto size" (and using an unsigned integer) instead of minus one as seems it would simplify this logic.

Reply:
Good point. I agree that  `MkFixedSizeCoinbaseTx(CMutableTransaction &tx, int64_t coinbaseSize)` would be better using an unsigned int. 
Imo <0 for signalling is good because it gives you a range of signals.
Lmk if you (others?) feel it is important to change it.